### PR TITLE
feat(elf): handle tail strings in symbol matching

### PR DESCRIPTION
**Pull request by geyslan**
### 1. Explain what the PR does

f2e02fdd9 **feat(elf): handle tail strings in symbol matching**

```
The ELF linker performs string table compression through two mechanisms:
1. Deduplicating identical strings
2. Sharing suffixes (tail strings) between different strings

This optimization is commonly applied by modern linkers (GNU ld, gold,
LLVM lld) to reduce binary size.

Real-world example from libc.so.6 .dynstr section:
  - "__libc_system" stored at offset 0x2de9
  - "system"        references offset 0x2df0 (0x2de9 + 7 bytes)

  Memory layout:
    0x2de9: __libc_system\0
                   └─ 0x2df0: system\0 (tail string)

The previous implementation did not account for these tail string
references, causing symbol lookups to fail when symbols used suffix-
sharing optimization.

Enhance findWantedSymbolNames() to detect both exact matches and tail
string matches when scanning the ELF string table:

1. Check for exact matches at string boundaries
2. Check for tail string (suffix) matches using strings.HasSuffix()
3. Prioritize exact matches over tail strings when both exist
4. Return only tail strings when no exact match is found for a symbol
5. Add comprehensive test suite with 13 test cases covering:
  * Exact matches vs tail string matches
  * Priority handling (exact matches preferred)
  * Multiple tail strings across different symbols
  * Order independence
  * Edge cases (empty tables, single characters, etc.)

Reference: https://docs.oracle.com/cd/E23824_01/html/819-0690/ggdlu.html
```

### CODE REVIEW

File: common/elf/symbols_test.go [hunk 1]
```diff
@@ -115,3 +118,210 @@ func TestErrSymbolNotFound(t *testing.T) {
...
+			strtab: []byte{
+				0,                // offset 0: empty string
+				'f', 'o', 'o', 0, // offset 1: "foo" (exact)
+				'b', 'a', 'r', '_', 'f', 'o', 'o', 0, // offset 5: "bar_foo" (tail is "foo")
+			},
+			wantedSymbols: []WantedSymbol{
+				PlainSymbolName("foo"),
+			},
+			expectedOffsets: []uint32{1, 9}, // "foo" at 1 (exact) and 9 (tail of "bar_foo")
+			expectedCount:   2,
```
[the remark this answers is not in the document]
Comment (geyslan) [reply common/elf/symbols_test.go]: This is a real bug. Good catch. Changed the implementation to solve it.

File: common/elf/symbols.go [hunk 2]
```diff
@@ -262,23 +280,48 @@ func findWantedSymbolNames(strtab []byte, wantedSymbols []WantedSymbol) (map[uin
...
+				for _, wantedSymbol := range wantedSymbols {
+					wantedStr := wantedSymbol.String()
+
+					// Check for exact match at this offset
+					if wantedSymbol.Matches(str) {
+						exactMatches[uint32(currString)] = struct{}{}
+						hasExactMatch[wantedStr] = struct{}{}
+					} else {
+						// Check for tail string match (suffix)
+						if len(str) > len(wantedStr) && strings.HasSuffix(str, wantedStr) {
```
[the remark this answers is not in the document]
Comment (geyslan) [reply common/elf/symbols.go:305]: If the entire string is the tail, it should be matched in the if branch above.

File: common/elf/symbols.go [hunk 3]
```diff
@@ -262,23 +280,48 @@ func findWantedSymbolNames(strtab []byte, wantedSymbols []WantedSymbol) (map[uin
...
+
+				// Check all wanted symbols - either exact match or tail string
+				for _, wantedSymbol := range wantedSymbols {
+					wantedStr := wantedSymbol.String()
+
+					// Check for exact match at this offset
+					if wantedSymbol.Matches(str) {
+						exactMatches[uint32(currString)] = struct{}{}
+						hasExactMatch[wantedStr] = struct{}{}
+					} else {
```
Comment (oshaked1) [inline common/elf/symbols.go]: Tail string matches are only relevant for `PlainSymbolName`, we should probably check that `wantedSymbol` is the correct type or add a `SubstringMatches` method to the interface that will return the matched offset if relevant.

Comment (geyslan) [reply common/elf/symbols.go]: As we discussed, I'm going to check the type.

Comment (geyslan) [reply common/elf/symbols.go]: Done, please LMK.

**Comment by oshaked1** [comment]
LGTM
